lib/spack/spack/fetch_strategy.py

```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
         self.submodules_delete = kwargs.get('submodules_delete', False)
         self.get_full_repo = kwargs.get('get_full_repo', False)
 
-    @property
-    def git_version(self):
-        vstring = self.git('--version', output=str).lstrip('git version ')
+        try:
+            self.ref = GitRef.from_directive(
+                commit=self.commit,
+                tag=self.tag,
+                branch=self.branch,
+                version_name=self.version_name)
+        except InvalidGitRef as e:
+            raise FetcherConflict(
+                'failed to identity any refspec (commit, tag, or branch) '
+                'for the git repo at {0}: {1}'.format(self.url, e))
+
+    @staticmethod
+    @memoized
+    def _get_git_version(git):
+        vstring = git('--version', output=str).lstrip('git version ')
         return Version(vstring)
 
-    @property
+    @property                                               # type: ignore[misc]
+    @memoized
     def git(self):
-        if not self._git:
-            self._git = which('git', required=True)
+        git = which('git', required=True)
 
-            # Disable advice for a quieter fetch
-            # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
-            if self.git_version >= Version('1.7.2'):
-                self._git.add_default_arg('-c')
-                self._git.add_default_arg('advice.detachedHead=false')
+        # Disable advice for a quieter fetch
+        # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
+        if self._get_git_version(git) >= Version('1.7.2'):
+            git.add_default_arg('-c')
+            git.add_default_arg('advice.detachedHead=false')
 
-            # If the user asked for insecure fetching, make that work
-            # with git as well.
-            if not spack.config.get('config:verify_ssl'):
-                self._git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
+        # If the user asked for insecure fetching, make that work
+        # with git as well.
+        if not spack.config.get('config:verify_ssl'):
+            git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
 
-        return self._git
+        return git
 
     @property
     def cachable(self):
-        return self.cache_enabled and bool(self.commit or self.tag)
+        return (self.cache_enabled and
+                not self.ref.always_check_for_updates_but_accept_airgapped())
```

Comment:
If we're going to use this in places not related to updating the variable, we need a better name for it.

Maybe something like `self.ref.is_mutable()` as an abstraction layer for `self.ref._ref_type == 'branch'` in case we change the behavior for tags or git adds a new type of mutable ref

Reply:
`.is_mutable()` sounds great! We can keep the lengthy explanation in the `._fetch_ref()` method's docstring.